wgpu-core/src/indirect_validation/validate_draw.wgsl

```diff
@@ -0,0 +1,82 @@
+override supports_indirect_first_instance: bool;
+
+struct MetadataEntry {
+    // bits 0..30 are an offset into `src`
+    // bit 31 signifies that we are validating an indexed draw
+    src_offset: u32,
+    // bits 0..30 are an offset into `dst`
+    // bit 30 is the most significant bit of `vertex_or_index_limit`
+    // bit 31 is the most significant bit of `instance_limit`
+    dst_offset: u32,
+    vertex_or_index_limit: u32,
+    instance_limit: u32,
+}
+
+struct MetadataRange {
+    start: u32,
+    count: u32,
+}
+var<push_constant> metadata_range: MetadataRange;
+
+@group(0) @binding(0)
+var<storage, read> metadata: array<MetadataEntry>;
+@group(1) @binding(0)
+var<storage, read> src: array<u32>;
+@group(2) @binding(0)
+var<storage, read_write> dst: array<u32>;
+
+@compute @workgroup_size(64)
+fn main(@builtin(global_invocation_id) global_invocation_id: vec3u) {
+    if global_invocation_id.x >= metadata_range.count { return; }
+
+    let metadata = metadata[metadata_range.start + global_invocation_id.x];
+    var failed = false;
+
+    let is_indexed = ((metadata.src_offset >> 31) & 1u) == 1u;
+    let src_base_offset = ((metadata.src_offset << 2) >> 2);
+    let dst_base_offset = ((metadata.dst_offset << 2) >> 2);
+
+    let first_vertex_or_index = src[src_base_offset + 2];
+    let vertex_or_index_count = src[src_base_offset + 0];
+
+    {
+        let can_overflow = ((metadata.dst_offset >> 30) & 1u) == 1u;
```

Comment:
this would be a bit clearer using `extractBits` - if you don't want to use it for performance reasons, maybe a free function?

Reply:
The backends add some validation code for `extractBits` to cap the offset and count. I'll add a function.